File: sdk/python3/nmidsdk/client/request.py
import struct
import nmidsdk.model.const as const

class Request:
    def __init__(self):
        self.data_type = 0
        self.data = bytearray()
        self.data_len = 0
        self.handle = ""
        self.handle_len = 0
        self.params_type = const.PARAMS_TYPE_MSGPACK
        self.params_handle_type = const.PARAMS_HANDLE_TYPE_ENCODE
        self.params_len = 0
        self.params = bytearray()
        self.ret = bytearray()
        self.ret_len = 0

    def content_pack(self, dataType: int, handle:str, params: bytearray):
        self.data_type = dataType
        self.handle = handle
        self.handle_len = len(handle)
        self.params = params
        self.params_len = len(params)
        self.data_len = const.UINT32_SIZE + const.UINT32_SIZE + const.UINT32_SIZE + self.handle_len + const.UINT32_SIZE + self.params_len
        
        content = bytearray(self.data_len)
        struct.pack_into('>I', content, 0, self.params_type)
        struct.pack_into('>I', content, const.UINT32_SIZE, self.params_handle_type)
        struct.pack_into('>I', content, const.UINT32_SIZE * 2, self.handle_len)
        content[const.UINT32_SIZE * 3:const.UINT32_SIZE * 3 + self.handle_len] = handle.encode()
        struct.pack_into('>I', content, const.UINT32_SIZE * 3 + self.handle_len, self.params_len)
        content[const.UINT32_SIZE * 4 + self.handle_len:const.UINT32_SIZE * 4 + self.handle_len + self.params_len] = params
        
        self.data = content
        return self.data, self.data_len

    def encode_pack(self):
        len = const.MIN_DATA_SIZE + self.data_len
        data = bytearray(len)

        struct.pack_into('>I', data, 0, const.CONN_TYPE_CLIENT)
        struct.pack_into('>I', data, const.UINT32_SIZE, self.data_type)
        struct.pack_into('>I', data, const.UINT32_SIZE * 2, self.data_len)
        data[const.MIN_DATA_SIZE:] = self.data

        return data

File: sdk/node/lib/client/request.js
'use strict';

// Client request encoder.
// Mirrors sdk/python3/nmidsdk/client/request.py and pkg/client/request.go.
//
// Packet layout (big-endian):
//   header (12 bytes): ConnType(4) | DataType(4) | DataLen(4)
//   body (DataLen bytes) for PDT_C_DO_JOB:
//     ParamsType(4) | ParamsHandleType(4) | HandleLen(4) | Handle |
//     ParamsLen(4) | Params

const C = require('../const');

class Request {
  constructor() {
    this.dataType = 0;
    this.data = Buffer.alloc(0);
    this.dataLen = 0;

    this.handle = '';
    this.handleLen = 0;
    this.paramsType = C.PARAMS_TYPE_MSGPACK;
    this.paramsHandleType = C.PARAMS_HANDLE_TYPE_ENCODE;
    this.paramsLen = 0;
    this.params = Buffer.alloc(0);

    this.ret = Buffer.alloc(0);
    this.retLen = 0;
  }

  // Build the body for a client do-job request.
  contentPack(dataType, handle, params) {
    this.dataType = dataType;
    this.handle = handle;
    this.handleLen = Buffer.byteLength(handle, 'utf8');
    this.params = params;
    this.paramsLen = params.length;
    this.dataLen =
      C.UINT32_SIZE + // paramsType
      C.UINT32_SIZE + // paramsHandleType
      C.UINT32_SIZE + // handleLen
      this.handleLen +
      C.UINT32_SIZE + // paramsLen
      this.paramsLen;

    const content = Buffer.alloc(this.dataLen);
    let offset = 0;
    content.writeUInt32BE(this.paramsType, offset);
    offset += C.UINT32_SIZE;
    content.writeUInt32BE(this.paramsHandleType, offset);
    offset += C.UINT32_SIZE;
    content.writeUInt32BE(this.handleLen, offset);
    offset += C.UINT32_SIZE;
    content.write(handle, offset, this.handleLen, 'utf8');
    offset += this.handleLen;
    content.writeUInt32BE(this.paramsLen, offset);
    offset += C.UINT32_SIZE;
    this.params.copy(content, offset);

    this.data = content;
    return content;
  }

  // Prepend the 12-byte header.
  encodePack() {
    const total = C.MIN_DATA_SIZE + this.dataLen;
    const data = Buffer.alloc(total);
    data.writeUInt32BE(C.CONN_TYPE_CLIENT, 0);
    data.writeUInt32BE(this.dataType, C.UINT32_SIZE);
    data.writeUInt32BE(this.dataLen, C.UINT32_SIZE * 2);
    this.data.copy(data, C.MIN_DATA_SIZE);
    return data;
  }
}

module.exports = Request;
